### DISCUSSION TITLE
Managing issues

### DISCUSSION THREAD

**Discussion by detrin** [shap/shap#3219] [General] [answered]
To make orientation in issues easier I have created the following repository
https://github.com/detrin/shap-issues

All open issues ordered by their id number https://htmlpreview.github.io/?https://github.com/detrin/shap-issues/blob/main/issues_v1_open.html
All open issues ordered by length ascending https://htmlpreview.github.io/?https://github.com/detrin/shap-issues/blob/main/issues_v1_open_ordered.html

Here is an example from first table
![image](https://github.com/shap/shap/assets/18104456/743f2cf2-cd0c-4b30-bf62-086514f2f328)
columns: description_quality, summary, short_label and bug are generated with GPT, if you are interested in prompts they are in the repository. It is all suggested, but I believe this might simplify orientation in issues.

P.S. the GPT is still working on the issues

**Comment by detrin** [comment]
I finished labelling the issues, the GPT4 part is maybe good for summarization, but the labels  are often incorrect. Anyway having basic info about issues helped me to orient myself in the issues. I want to contribute further sorting those issues and distilling what is relevant for the future of shap. I believe that classifying the issues and closing stale ones will help the project to stay longer alive.
When I went over many issues this weekend I learned new things, sometimes the ideas are interesting, but most of the time it is noise.

What is my plan regarding issues:

1. I would proceed with pinging ppl via my account with template messages on issues that were not updated in let's say one year (with github API). 
2. The rest would be inspected manually to assign label.
3. After let's say a week of no response I would again with github API close those issues that were notified and label them as stale. The rest where ppl responded will have to be manually checked and labelled. I am sure there will be plenty of issues where ppl respond, but it needs more clarification.

What would be the benefit?

1. As I mentioned, we can extract what is actually important for the package and quantify the issues with labels this way.
2. Reducing the issues would also make shap more attractive and will help with orientation in issues. Having issues as backlog is not that bad idea.
3. I believe this might attract more ppl to contribute to this package.

Why it would be bad idea?

1. One might argue that it is a waste of time if the most of (relevant) issues won't be closed anyway.
2. The time could be used for something more productive in shap if we consider the expected lifetime of this package.
3. Pinging users might create some bad PR (I am just guessing)

Those are my ideas, please have a look at example [table](https://htmlpreview.github.io/?https://github.com/detrin/shap-issues/blob/main/issues_v1_open_ordered_c.html) and optionally the [repo](https://github.com/detrin/shap-issues). Any ideas or suggestions are welcomed.

@thatlittleboy @dsgibbons @connortann @Z4K4RI4

**Reply by thatlittleboy** [reply, marked as answer]
@detrin Thanks for this. I agree in principal about the benefits of doing so, some of the arguments were already brought up in https://github.com/shap/shap/discussions/3051 so I won't rehash them here.

A couple of points from me

1. Is there any fundamental difference between this v.s. using of the available stale bots on Github workplace? E.g. [stale](https://github.com/actions/stale) or [close-stale-issues](https://github.com/marketplace/actions/close-stale-issues#stale-issue-message) as you suggested in the other discussion.
2. > I would proceed with pinging ppl via my account with template messages on issues that were not updated

    If the messages are generated via GPT, please make it clear that it is so (**including** the ones that you have already sent out, if any). Especially if the messages are generating suggestions to the user; it might be potentially misleading to future readers.
3. I still feel 1 week from stale->close is too short (for a bot-initiated closure). Connor and I agreed on (2 years, 3 months) over at https://github.com/shap/shap/discussions/3051 rather than (1 year, 1 week).

**Reply by detrin** [reply]
@thatlittleboy 
Ad 1. I don't think so, on one hand replying with real account might look more humane and can be done in semi-automated way.
Ad2. I wouldn't like to get GPT response on my issue that is really old. I think human-human interaction is important even at this level. My plan was to use keywords in summarization prepared by GPT or keywords in the body for segmenting the issues. Each segment would have a certain template message that would be written by my hands. I think it might be helpful to use generated labels for real labelling or as a draft. However, when the error rate is high it doesn't make sense to use it.
Ad 3. Two years is more reasonable. I think having one month for a reply should be sufficient, I think if ppl do not opt out they will get a notification to email. So unless one is really not using github at all, they will notice the new comment on their issue. 

Next I will try to find the right segments and then start testing it on batch of for example 100 issues. A test of sending comments will be done on my created issues. I just hope they won't ban me for spam hah.

I was also considering posting the scraped comments to kaggle, maybe someone would come up with something clever for classification in short time. I am not sure whether it would be possible due to rights even though comments are publicly available.

**Reply by connortann** [reply]
It's interesting to see your use of GPT to analyse those messages - thanks for looking in to it and sharing that repo. 

I agree with @thatlittleboy 's preference to use the GitHub bot where possible to automate these tasks, rather than creating our own automation. The GitHub stale bot is fairly well established and understood by the GitHub community, and using an existing tool means one less thing for us to document and maintain.

If we do use any automation in messages, I much prefer doing it in a transparent manner from a Bot account, rather than generating messages that superficially appear to be personal. So, I'd be opposed to using GPT to send messages from a personal account.

Let's use the thread on #3051 to discuss the implementation details of the stale bot.
